File: examples/cpp/banana/EvaluateBanana.cpp
/**
This example shows how to construct a map from \f$\mathbb{R}^2\rightarrow \mathbb{R}\f$  with the form.
 \f[
T_2(x_1,x_2) = x_1^2 + x_2
 \f] 
 Note that this is one component of the map used to define a "banana'"
 */

#include <MParT/ConditionalMapBase.h>
#include <MParT/MapFactory.h>
#include <MParT/MultiIndices/MultiIndexSet.h>
#include <MParT/Utilities/ArrayConversions.h>

using namespace mpart; 
 
int main(int argc, char* argv[]){

    Kokkos::initialize(argc,argv);
    {

    unsigned int dim = 2;

    Eigen::MatrixXi multis(2,dim);
    multis << 0,1,
              2,0;

    MultiIndexSet mset(multis);

    MapOptions opts;
    opts.basisType    = BasisTypes::ProbabilistHermite;
    opts.posFuncType = PosFuncTypes::SoftPlus;
    opts.quadType    = QuadTypes::AdaptiveSimpson;
    opts.quadAbsTol  = 1e-6;
    opts.quadRelTol  = 1e-6;

    std::shared_ptr<ConditionalMapBase> map = MapFactory::CreateComponent(mset.Fix(), opts);

    unsigned int numCoeffs = mset.Size();
    Eigen::VectorXd coeffs(numCoeffs);
    map->SetCoeffs(coeffs); 

    coeffs(0) = 1.0; // Set the linear coefficient 
    coeffs(1) = 0.5; // Set the quadratic coefficient


    unsigned int numPts = 128;
    Eigen::RowMatrixXd pts = Eigen::RowMatrixXd::Random(dim,numPts);
    Eigen::RowMatrixXd evals = map->Evaluate(pts);
    Eigen::VectorXd logDet = map->LogDeterminant(pts);

    std::cout << "Map Evaluations:\n" << evals << std::endl;
    }
    Kokkos::finalize();

    return 0;
}

File: examples/python/banana/EvaluateBanana.py
from mpart import *
import numpy as np
import matplotlib.pyplot as plt

# Create multi-index set:
multis = np.array([[0, 1], [2, 0]])
mset = MultiIndexSet(multis)

fixed_mset = mset.fix(True)

# Set MapOptions
opts = MapOptions()

# opts.basisType   = BasisTypes.ProbabilistHermite
# opts.posFuncType = PosFuncTypes.SoftPlus
# opts.quadType    = QuadTypes.AdaptiveSimpson
# opts.quadAbsTol  = 1e-6
# opts.quadRelTol  = 1e-6

map = CreateComponent(fixed_mset, opts)
# num_coeffs = mset.Size()  # not needed in python example but works

coeffs = np.array([1.0, 0.5]) 
# coeffs = None
map.SetCoeffs(coeffs)

print('Coeffs set')
print(map.CoeffMap())
print('============')

pts = np.random.randn(2,5)
print('The points')
print(pts)
print('============')

map_of_pts = map.Evaluate(pts) 
print('Map of the points')
print(map_of_pts)
print('============')

log_det = map.LogDeterminant(pts)
print('Log det of map at the points')
print(log_det)
print('============')


# Ways to change coeffs of map
# coeffs = np.array([2.0, 0.5]) # map does NOT change
# coeffs[:] = np.array([2.0, 0.5]) # map does change (maybe we don't want it to)
# coeffs[0] = 2 # map does change (maybe we don't want it to)
# map.CoeffMap()[:] = np.array([2, 0.5]) # map does change (we want it to)

coeffs[:] = np.array([4, 0.5])  # changes map, we want this to NOT change the map

map_of_pts = map.Evaluate(pts)
print('Map of the points (coeffs changed)')
print(map_of_pts)
print('============')



